### PR TITLE
Export panels in TXT format from the gene panel page

### LINKED ISSUE #3822: Export gene panels in a scout loadable csv form from WUI
Currently we do pdf exports and html table from WUI, but have csv export from CLI. It seems of interest for users moving between institutes (or sites) to have easy access to the latest csv version. We could also allow "copying" of panels between institutes, but just dl of csv solves the issue without having to worry about ownership, updates etc.

### PR DESCRIPTION
**Pull request by northwestwitch**
Fix #3822

image

<details>
<summary>Testing on cg-vm1 server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. Make sure the PR is pushed and available on [Docker Hub](https://hub.docker.com/repository/docker/clinicalgenomics/scout-server-stage)
1. Fist book your testing time using the Pax software available at [https://pax.scilifelab.se/](https://pax.scilifelab.se). The resource you are going to call dibs on is `scout-stage` and the server is `cg-vm1`.
1. `ssh <USER.NAME>@cg-vm1.scilifelab.se`
1. `sudo -iu hiseq.clinical`
1. `ssh localhost`
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `podman ps`
1. Stop the service with current deployed branch: `systemctl --user stop scout.target`
1. Start the scout service with the branch to test: `systemctl --user start scout@<this_branch>`
1. Make sure the branch is deployed: `systemctl --user status scout.target`
1. After testing is done, repeat procedure at [https://pax.scilifelab.se/](https://pax.scilifelab.se), which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

<details>
<summary>Testing on hasta server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. `ssh <USER.NAME>@hasta.scilifelab.se`
1. Book your testing time using the Pax software. `us; paxa -u <user> -s hasta -r scout-stage`. You can also use the WSGI Pax app available at [https://pax.scilifelab.se/](https://pax.scilifelab.se).
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `conda activate S_scout; pip freeze | grep scout-browser`
1. Deploy the branch to test: `bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_scout -t scout -b <this_branch>`
1. Make sure the branch is deployed: `us; scout --version`
1. After testing is done, repeat the `paxa` procedure, which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

**How to test**:
1. On a demo local instance, go to demo gene panel and download it in text file using the button shown in the pic abobve

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [x] code approved by DN
- [x] tests executed by CR

### CODE REVIEW

File: scout/server/blueprints/panels/controllers.py [hunk 1]
```diff
@@ -305,7 +306,26 @@ def new_panel(
     return panel_id
 
 
-def panel_export(store, panel_obj):
+def downloaded_panel_name(panel_obj, format) -> str:
```
Comment (northwestwitch) [inline scout/server/blueprints/panels/controllers.py:309]: New function

File: scout/server/blueprints/panels/views.py [hunk 2]
```diff
@@ -252,23 +254,39 @@ def panel_restore(panel_id):
     return redirect(url_for("panels.panels", panel_id=panel_obj["_id"]))
 
 
-@panels_bp.route("/panels/export-panel/<panel_id>", methods=["GET", "POST"])
-def panel_export(panel_id):
+@panels_bp.route("/panels/export-panel-txt/<panel_id>", methods=["GET"])
+def panel_export_txt(panel_id):
```
Comment (northwestwitch) [inline scout/server/blueprints/panels/views.py:258]: New function

File: scout/server/blueprints/panels/views.py [hunk 3]
```diff
@@ -252,23 +254,39 @@ def panel_restore(panel_id):
...
+
+    return Response(
+        generate(),
+        mimetype="text/txt",
+        headers={"Content-Disposition": f"attachment;filename={file_name}"},
+    )
+
+
+@panels_bp.route("/panels/export-panel-pdf/<panel_id>", methods=["GET"])
+def panel_export_pdf(panel_id):
```
Comment (northwestwitch) [inline scout/server/blueprints/panels/views.py:281]: Old function, renamed

File: tests/server/blueprints/panels/test_panels_views.py [hunk 4]
```diff
@@ -143,33 +143,29 @@ def test_panels(app, institute_obj):
...
-    resp = client.get(url_for("panels.panel_export", panel_id=panel_obj["_id"]))
-    # THEN it should display the panel with all the genes
+    # WHEN using the panels.panel_export_txt endpoint
+    resp = client.get(url_for("panels.panel_export_txt", panel_id=panel_obj["_id"]))
+    # THEN it should download the panel in the right format
     assert resp.status_code == 200
+    assert resp.mimetype == "text/txt"
 
 
-def test_panel_export_case_hits(client, real_panel_database):
```
Comment (northwestwitch) [inline tests/server/blueprints/panels/test_panels_views.py:158]: This test was testing the wrong endpoint (see #3847), should be rewritten, but we can do it in another PR

File: tests/server/blueprints/panels/test_panels_views.py [hunk 5]
```diff
@@ -143,33 +143,29 @@ def test_panels(app, institute_obj):
         assert resp.status_code == 200
 
 
-# This test is slow since pdf rendering is slow
-def test_panel_export(client, real_panel_database):
+def test_gene_panel_export_txt(client, real_panel_database):
```
Comment (northwestwitch) [inline tests/server/blueprints/panels/test_panels_views.py:146]: New test

File: scout/server/blueprints/panels/views.py [hunk 6]
```diff
@@ -310,7 +327,7 @@ def panel_export_case_hits(panel_id):
             str(data["panel"]["version"]),
             institute_id,
             case_name,
-            datetime.datetime.now().strftime(DATETIME_FORMATTER),
+            datetime.datetime.now().strftime(DATE_DAY_FORMATTER),
```
Comment (dnil) [inline scout/server/blueprints/panels/views.py:330]: Why not just reuse the stored `now`?

File: scout/commands/view/panels.py [hunk 7]
```diff
@@ -3,6 +3,7 @@
 import click
 from flask.cli import with_appcontext
 
+from scout.constants import DATE_DAY_FORMATTER
```
Comment (dnil) [inline scout/commands/view/panels.py:6]: 💯

File: scout/server/blueprints/panels/views.py [hunk 8]
```diff
@@ -32,7 +34,6 @@
 from .forms import PanelGeneForm
 
 LOG = logging.getLogger(__name__)
-DATETIME_FORMATTER = "%Y-%m-%d"
```
Comment (dnil) [inline scout/server/blueprints/panels/views.py:35]: 👍

File: scout/server/blueprints/panels/controllers.py (hunk 1 above)
Comment (dnil) [inline scout/server/blueprints/panels/controllers.py:309]: Also optionally institute and case for the panel extent report! As is the extent report pdf and standard table pdf collide.

File: scout/server/blueprints/panels/templates/panels/panel.html [hunk 9]
```diff
@@ -105,7 +105,8 @@
       {% endif %}
       <li class="list-group-item">
         <div class="d-flex justify-content-between">
-          <a href="{{ url_for('panels.panel_export', panel_id=panel._id) }}" class="btn btn-secondary btn-sm" download>Export to PDF</a>
+          <a href="{{ url_for('panels.panel_export_pdf', panel_id=panel._id) }}" class="btn btn-secondary btn-sm" download>Export to PDF</a>
+          <a href="{{ url_for('panels.panel_export_txt', panel_id=panel._id) }}" class="btn btn-secondary btn-sm" download>Export to TXT</a>
```
Comment (dnil) [inline scout/server/blueprints/panels/templates/panels/panel.html:109]: A little tooltip bragging that exported TXT panels are in a format that can be imported into other institutes or Scout instances perhaps?

File: scout/server/blueprints/panels/views.py [hunk 10]
```diff
@@ -287,7 +304,7 @@ def panel_export_case_hits(panel_id):
         institute_id, case_name = request.form.get("case_name").strip().split(" - ")
     except ValueError:
         flash(
-            "Could not parse case name, plase use format: 'cust000 - 643594' or use typing suggestions",
+            "Could not parse case name, please use format: 'cust000 - 643594' or use typing suggestions",
```
Comment (dnil) [inline scout/server/blueprints/panels/views.py:307]: 😄

**Review by dnil — APPROVED** [review]
Works like a charm! Some minor suggestions in comments.

File: scout/server/blueprints/panels/views.py (hunk 6 above)
Comment (northwestwitch) [reply scout/server/blueprints/panels/views.py:330]: Which now? I wanted something that had just year month and day, not the whole thing with minutes seconds etc

File: scout/server/blueprints/panels/controllers.py (hunk 1 above)
Comment (northwestwitch) [reply scout/server/blueprints/panels/controllers.py:309]: Ah yes, good suggestion!

File: scout/server/blueprints/panels/templates/panels/panel.html (hunk 9 above)
Comment (northwestwitch) [reply scout/server/blueprints/panels/templates/panels/panel.html:109]: Yes, that would be nice!

File: scout/server/blueprints/panels/views.py [hunk 11]
```diff
@@ -300,7 +317,7 @@ def panel_export_case_hits(panel_id):
         )
         return redirect(request.referrer)
     data = controllers.panel_export_case_hits(panel_id, institute_obj, case_obj)
-    now = datetime.datetime.now().strftime(DATETIME_FORMATTER)
+    now = datetime.datetime.now().strftime(DATE_DAY_FORMATTER)
```
Comment (dnil) [inline scout/server/blueprints/panels/views.py:320]: Here: you store `now` for the `report_created_at`. Not that it matters to the users when microseconds are not shown. And you are perhaps going to use the filename function for it as well!

File: scout/server/blueprints/panels/views.py (hunk 6 above)
Comment (dnil) [reply scout/server/blueprints/panels/views.py:330]: A few lines up..

File: scout/server/blueprints/panels/views.py (hunk 11 above)
Comment (northwestwitch) [reply scout/server/blueprints/panels/views.py:320]: Ahh yes, but it's a very log and user-unfriendly string I think

Comment (northwestwitch) [reply scout/server/blueprints/panels/views.py:320]: Ah yes

File: scout/server/blueprints/panels/controllers.py (hunk 1 above)
Comment (northwestwitch) [reply scout/server/blueprints/panels/controllers.py:309]: What did you mean for` panel extent report`? I assumed it was this endpoint: https://github.com/Clinical-Genomics/scout/blob/ef98404273e180d7a6d73ff40d1dccf823a691b6/scout/server/blueprints/panels/views.py#L281 but it already contains institute and case name? 🤔

File: scout/server/blueprints/panels/templates/panels/panel.html (hunk 9 above)
Comment (northwestwitch) [reply scout/server/blueprints/panels/templates/panels/panel.html]: Done: 
image
